projects/newprofile_ping_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,518 @@
+```
+Try to pair each `new-profile` ping with reason `shutdown` to the very first `main` ping with reason `shutdown` received from that client, to make sure that the former were sent at the right time.
+
+
+```python
+first_main = main_deduped.filter(lambda p:\
+                                    p.get("payload/info/previousSessionId") == None and\
+                                    p.get("payload/info/reason") == "shutdown")
+```
+
+```python
+newping_shutdown = deduped_docid.filter(lambda p: p.get("payload/reason") == "shutdown")
+```
+
+```python
+newprofile_plus_main = first_main.union(newping_shutdown)
+sorted_per_client = newprofile_plus_main.map(lambda p: (p["clientId"], [(p, p["meta/creationTimestamp"])]))\
+                                        .reduceByKey(lambda a, b: sorted(a + b, key=lambda k: k[1]))\
+                                        .filter(lambda p: len(p[1]) > 1)\
+                                        .map(lambda p: (p[0], [r[0] for r in p[1]]))
+num_analysed_clients = sorted_per_client.count()
+```
+
+```python
+HALF_HOUR_IN_S = 30 * 60
+
+def is_newprofile(p):
+    # The 'main' ping has the reason field in 'payload/info/reason'
+    return "payload/reason" in p and p.get("payload/reason") in ["startup", "shutdown"]
+
+def validate_newprofile_scheduling(client_data):
+    ordered_pings = client_data[1]
+    # If there's a new-profile ping with reason 'shutdown', look for the closest next
+    # 'main' ping with reason shutdown.
+    for i, p in enumerate(ordered_pings):
+        # Skip until we find the 'new-profile' ping.
+        if not is_newprofile(p):
+            continue
+            
+        # We found the 'new-profile' ping. Do we have any other ping
+        # after this?
+        next_index = i + 1
+        if next_index >= len(ordered_pings):
+            return ("No more pings after the 'new-profile'", 1)
+    
+        # Is the next ping a 'main' ping with reason 'shutdown'?
+        next_ping = ordered_pings[next_index]
+        if is_newprofile(next_ping):
+            return ("Duplicate 'new-profile' ping.", 1)
+        
+        if next_ping.get("payload/info/reason") == "shutdown":
+            if next_ping.get("payload/info/sessionLength") <= HALF_HOUR_IN_S:
+                return ("The 'new-profile' ping was correctly scheduled", 1)
+            
+            return ("The 'new-profile' ping was scheduled at the wrong time", 1)
+            
+    return ("No 'shutdown' new-profile ping found", 1)
+
+scheduling_error_counts = sorted_per_client.map(validate_newprofile_scheduling).countByKey()
+```
+
+```python
+for error, count in scheduling_error_counts.items():
+    print("{}:\t{}".format(error, pct(count, num_analysed_clients)))
+```
+    The 'new-profile' ping was correctly scheduled:	98.3717774763
+    The 'new-profile' ping was scheduled at the wrong time:	0.881953867028
+    Duplicate 'new-profile' ping.:	0.542740841248
+    No 'shutdown' new-profile ping found:	0.203527815468
+
```

Comment:
Missing validation case: we didn't check that new-profile/startup is sent only from sessions that last longer than 30min

I don't know that we absolutely must validate that case. Up to you.